### ISSUE TITLE
Shows "VPC should have at least 2 associated subnets in different availability zones" but my VPC has 2 subnets in different AZ's

### ISSUE THREAD

**Issue by mobereng** [docker/compose-cli#864]
I tried to use the option x-aws-vpc on my docker compose file, so I can specify an existing VPC, but as long as I run docker compose up it shows "VPC should have at least 2 associated subnets in different availability zones" which is wrong cause I have 2 different subnets on different AZ's on that VPC, so I'm not sure if there's a missconfiguration on my VPC or someting that I missed on the compose file.

**Steps to reproduce the issue:**
1.
3.
2.

**Describe the results you received:**
VPC should have at least 2 associated subnets in different availability zones

**Describe the results you expected:**
Resources to be created on the VPC i specify

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Cloud integration: 1.0.1
 Version:           19.03.6-ce
 API version:       1.40
 Go version:        go1.13.4
 Git commit:        369ce74
 Built:             Fri May 29 04:01:26 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          19.03.6-ce
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.4
  Git commit:       369ce74
  Built:            Fri May 29 04:01:57 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.2
  GitCommit:        ff48f57fc83a8c44cf4ad5d672424a98ba37ded6
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker context show`:**  
You can also run `docker context inspect context-name` to give us more details but don't forget to remove sensitive content.

```
[
    {
        "Name": "test",
        "Metadata": {
            "Description": "(eu-west-2)"
        },
        "Endpoints": {
            "docker": {
                "SkipTLSVerify": false
            },
            "ecs": {
                "Profile": "test",
                "Region": "eu-west-2"
            }
        },
        "TLSMaterial": {},
        "Storage": {
            "MetadataPath": "/path",
            "TLSPath": "/path"
        }
    }
]
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false

Server:
 Containers: 1
  Running: 0
  Paused: 0
  Stopped: 1
 Images: 2
 Server Version: 19.03.6-ce
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ff48f57fc83a8c44cf4ad5d672424a98ba37ded6
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.14.198-152.320.amzn2.x86_64
 Operating System: Amazon Linux 2
 OSType: linux
 Architecture: x86_64
 CPUs: 1
 Total Memory: 983.3MiB
 Name: ip-10-0-0-180.eu-west-2.compute.internal
 ID: UGZR:7DAU:VJDC:E2JB:Z5P2:FI3Q:ST4K:HAUU:SGJ2:HQSU:ZEGS:UKRT
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS ECS, Azure ACI, local, etc.):**

**Comment by ndeloof** [comment]
Can you please install the [AWS CLI](https://aws.amazon.com/fr/cli/) and run `aws --region eu-west-3 ec2 describe-subnets --filter Name=vpc-id,Values=vpc-xxx` to confirm the API do actualy list the expected subnets for your VPC and this is an ECS integration bug?

**Comment by mobereng** [comment]
For sure, it shows the following (sensitive info removed)

`{
    "Subnets": [
        {
            "AvailabilityZone": "eu-west-2a",
            "AvailabilityZoneId": "euw2-az2",
            "AvailableIpAddressCount": 249,
            "CidrBlock": "10.0.0.0/24",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": false,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-xxx",
            "VpcId": "vpc-xx",
            "OwnerId": "xxx",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "Tags": [
                {
                    "Key": "Environment",
                    "Value": "master"
                },
                {
                    "Key": "Name",
                    "Value": "Public_Subnet_A"
                }
            ],
            "SubnetArn": "arn:xxx"
        },
        {
            "AvailabilityZone": "eu-west-2b",
            "AvailabilityZoneId": "euw2-az3",
            "AvailableIpAddressCount": 251,
            "CidrBlock": "10.0.3.0/24",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": false,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-xxx",
            "VpcId": "vpc-xxx",
            "OwnerId": "xxx",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "Tags": [
                {
                    "Key": "Name",
                    "Value": "Private_Subnet_B"
                },
                {
                    "Key": "Environment",
                    "Value": "master"
                }
            ],
            "SubnetArn": "arn:xxx"
        },
        {
            "AvailabilityZone": "eu-west-2c",
            "AvailabilityZoneId": "euw2-az1",
            "AvailableIpAddressCount": 251,
            "CidrBlock": "10.0.5.0/24",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": false,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-xxx",
            "VpcId": "vpc-xxx",
            "OwnerId": "xxx",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "Tags": [
                {
                    "Key": "Name",
                    "Value": "Private_Subnet_C"
                }
            ],
            "SubnetArn": "arn:xxx"
        },
        {
            "AvailabilityZone": "eu-west-2b",
            "AvailabilityZoneId": "euw2-az3",
            "AvailableIpAddressCount": 250,
            "CidrBlock": "10.0.2.0/24",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": false,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-xxx",
            "VpcId": "vpc-xxx",
            "OwnerId": "xxx",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "Tags": [
                {
                    "Key": "Name",
                    "Value": "Public_Subnet_B"
                },
                {
                    "Key": "Environment",
                    "Value": "master"
                }
            ],
            "SubnetArn": "arn:xxx"
        },
        {
            "AvailabilityZone": "eu-west-2a",
            "AvailabilityZoneId": "euw2-az2",
            "AvailableIpAddressCount": 251,
            "CidrBlock": "10.0.1.0/24",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": false,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-xxx",
            "VpcId": "vpc-xxx",
            "OwnerId": "xxx",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "Tags": [
                {
                    "Key": "Name",
                    "Value": "Private_Subnet_A"
                },
                {
                    "Key": "Environment",
                    "Value": "master"
                }
            ],
            "SubnetArn": "arn:xxx"
        },
        {
            "AvailabilityZone": "eu-west-2c",
            "AvailabilityZoneId": "euw2-az1",
            "AvailableIpAddressCount": 251,
            "CidrBlock": "10.0.4.0/24",
            "DefaultForAz": false,
            "MapPublicIpOnLaunch": false,
            "MapCustomerOwnedIpOnLaunch": false,
            "State": "available",
            "SubnetId": "subnet-xxx",
            "VpcId": "vpc-xxx",
            "OwnerId": "xxx",
            "AssignIpv6AddressOnCreation": false,
            "Ipv6CidrBlockAssociationSet": [],
            "Tags": [
                {
                    "Key": "Name",
                    "Value": "Public_Subnet_C"
                }
            ],
            "SubnetArn": "arn:xxx"
        }
    ]
}`

**Comment by mobereng** [comment]
Also, is there any way where I can specify which subnets use ? cause if there is a private layer service I need it to run over the private subnets but if it is maybe an nginx I need that one on public subnets, so I'm a bit confused about how to specify that on the yml file

**Comment by ndeloof** [comment]
Not sure why we don't retrieve VPC's subnets. I've created https://github.com/docker/compose-cli/pull/866 as a possible fix, maybe in some circumstances the API only return < 2 subnets on first API results page.

about defining subnets a service should be deployed to, this is something we don't support yet. Would need to create custom `network` extensions for this purpose.

**Comment by flaviostutz** [comment]
I had this same error last week and in my case the cause was I was using a region that was "classic EC2", so that there was no default VPC in the region. I had to remove all elements that were using those "classic" elements and ask AWS to migrate my region manually.

**Comment by flaviostutz** [comment]
Please check if your problem is fixed by #852

**Comment by mobereng** [comment]
Still same issue @ndeloof @flaviostutz I already change my region to us-east-1 and migrate all my resources to there with 6 AZ's (us-east-1a to us-east-1d), still showing same message "VPC  should have at least 2 associated subnets in different availability zones"

**Comment by ndeloof** [comment]
@mobereng can you please build CLI from repository HEAD (`make` will generate `./bin/docker` to set in your PATH)
then run command with `-D` (debug) flag, so we can track API results from https://github.com/docker/compose-cli/blob/main/ecs/sdk.go#L191

**Comment by mobereng** [comment]
@ndeloof Already did, this is my current environment:

```
Client: Docker Engine - Community
 Cloud integration: 1.0.1-30-gdd5f66f
 Version:           19.03.13
 API version:       1.40
 Go version:        go1.13.15
 Git commit:        4484c46d9d
 Built:             Wed Sep 16 17:02:52 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.13
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       4484c46d9d
  Built:            Wed Sep 16 17:01:20 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.7
  GitCommit:        8fba4e9a7d01810a393d5d25a3621dc101981175
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

The deployment worked with respect to the "VPC should have at least 2 associated subnets in different availability zones" issue, but now does not seem to raise my services right, I will call this one solved, and review and raise another issue. But pretty much CloudMap does not register the service names so the containers end up being deleted, and also the volumes creation is failing it tries to register the same efs mount target per each AZ's and end up crashing as well. I will keep you posted, thanks for the help !

Is there any slack channel or another communication place appart from this github issues ?

**Comment by ndeloof** [comment]
we can discuss on https://dockercommunity.slack.com/archives/G016UBBRAE5 (EU timezone)

**Comment by zeknox** [comment]
Running into the same issue here.  2 subnets are visible when running the following command in different availability zones (us-east-1a, us-east-1b)

    aws --region us-east-1 ec2 describe-subnets --filter Name=vpc-id,Values=vpc-xxxx

**Comment by ndeloof** [comment]
I'm stuck here, can't see why using golang SDK's doesn't return the same values as the CLI
```golang
DescribeSubnetsWithContext(ctx, &ec2.DescribeSubnetsInput{
	Filters: []*ec2.Filter{
		{
			Name:   aws.String("vpc-id"),
			Values: []*string{aws.String(vpcID)},
		},
	},
```

**Comment by flaviostutz** [comment]
Just a tip of something got me these days... I had a problem with using
different regions both in "default" AWS CLI and in Golang SDK, even passing
the right region as a parameter to it.
I was trying to list the LB from region us-west-2 (that really existed in
AWS Console) but it always returned an empty response, even I was passing
"us-west-1" to the SDK.
Then I realized my default AWS CLI region was us-east-1 (which had no LBs).
I changed the AWS CLI default region to "us-west-2" and everything started
working.
Maybe you are experiencing a similar issue with Golang's SDK?

**Comment by gillegra** [comment]
I was experiencing something very similar to this. `docker compose up` kept failing because `A load balancer cannot be attached to multiple subnets in the same Availability Zone`. This was especially frustrating because I was able to launch this same stack the other day from a different machine without any trouble.

After some investigation I figured out that `docker compose up` was trying to use us-east-1, even though I configured my default region to us-east-2. When I created the ecs context, I chose to supply the AWS creds rather than use an existing profile. I didn't notice it at the time, but the region I chose, us-east-2, was put under `[profile default]` in ~/.aws/config. Turns out this needed to be under `[default]`. Once I fixed that, everything built correctly like before. Apparently compose is following the default config while context is setting the profile default.
